test/cpp/api/autograd.cpp

```diff
@@ -0,0 +1,129 @@
+      variable_list output = {grad_output[0] + grad_output[0]*var2, Variable(), grad_output[0] * mul + grad_output[0] * var1};
+      return output;
+    }
+  };
+
+  Variable x = torch::randn({5,5}, torch::requires_grad());
+  Variable y = torch::randn({5,5}, torch::requires_grad());
+  auto res = MyFunction::apply(x,2,y)[0];
+  auto go = torch::ones({}, torch::requires_grad());
+  res.sum().backward(go, false, true);
+
+  ASSERT_VARIABLE_EQ(x.grad(), y + torch::ones({5,5}));
+  ASSERT_VARIABLE_EQ(y.grad(), x + torch::ones({5,5})*2);
+}
+
+TEST(CustomAutogradTest, FunctionReturnsInput) {
+  struct MyFunction : public torch::autograd::Function<MyFunction> {
+    static variable_list forward(AutogradContext *ctx, Variable var1) {
+      return {var1};
+    }
+
+    static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
+      return {grad_output[0]*2};
+    }
+  };
+
+  Variable x(torch::ones(1, torch::requires_grad()));
+  MyFunction::apply(x)[0].backward(torch::ones(1) , true, true);
+  ASSERT_VARIABLE_EQ(x.grad(), torch::full(1,2));
+}
+
+TEST(CustomAutogradTest, NoGradCustomFunction) {
+  // Custom Function should respect grad mode
+ struct MyOp : public torch::autograd::Function<MyOp> {
+   static variable_list forward(AutogradContext *ctx, Variable x) {
+     return {x+1};
+   }
+
+   static variable_list backward(AutogradContext *ctx, variable_list dy) {
+     return dy;
+   }
+ };
+
+ auto x = torch::ones({5,5}, torch::requires_grad());
+ {
+    at::NoGradGuard no_grad;
+    auto y = MyOp::apply(x)[0];
+    ASSERT_FALSE(y.requires_grad());
+ }
+}
+
+TEST(CustomAutogradTest, MarkNonDifferentiable) {
+  struct MyFunction : public torch::autograd::Function<MyFunction> {
+    static variable_list forward(AutogradContext *ctx, Variable v) {
+      Variable output = v > 0;
+      ctx->mark_non_differentiable({output});
+      return {output};
+    }
+
+    static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
+      return { (grad_output[0]*0.0) };
+    }
+  };
+
+  auto x = torch::randn({5,5}, torch::requires_grad());
+  auto mask = MyFunction::apply(x)[0];
+  ASSERT_FALSE(mask.requires_grad());
+  auto y = x.masked_fill(mask, 0);
+  y.sum().backward();
+}
+
+TEST(CustomAutogradTest, ReturnLeafInplace) {
+  struct Inplace : public torch::autograd::Function<Inplace> {
+    static variable_list forward(AutogradContext *ctx, Variable a, Variable b) {
+      ctx->mark_dirty({a});
+      return {a.add_(b), b+2};
+    }
+
+    static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
+      return {grad_output[0], grad_output[0] + grad_output[1]};
+    }
+  };
+
+  Variable x = torch::randn({5,5});
+  Variable y = torch::randn({5,5}, torch::requires_grad());
+
+  auto out = Inplace::apply(x,y);
+  auto &q = out[0];
+  ASSERT_TRUE(torch::equal(q, x));
+  ASSERT_TRUE(q.requires_grad());
+  q.sum().backward();
+  ASSERT_VARIABLE_EQ(y.grad(), torch::ones({5,5}));
+}
```

Comment:
Should we also add a test case for returning a null gradient in backward(), similar to test_hook_none in test_autograd.py?

Reply:
The first test (CustomFunction) has a backward function that returns an empty Variable, which is treated the same way as "null" in the python code. Does that count? 
Also, I'll be adding all the other api tests from test_autograd.py in future PR.

Reply:
I see, I think since `variable_list` only accepts Variables (and can't convert `null`), using `Variable()` is a good idea.